### ISSUE TITLE
Unable to import shiboken2 with cx_freeze executable

### ISSUE THREAD

**Issue by Lois31** [marcelotduarte/cx_Freeze#767]
Hello Team,

It's been several days that I block on a problem between PySide2 and cx_freeze...
When I freeze my application, I get the following error when launching the executable:

> test-gui.exe
PySide2/__init__.py: Unable to import shiboken2 from L:\dev\PROJECT\bin\windows, L:\dev\PROJECT\bin\windows\lib\library.zip, L:\dev\PROJECT\bin\windows\lib
Traceback (most recent call last):
  File "I:\REF\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\__startup__.py", line 40, in run
    module.run()
  File "I:\REF\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\Console.py", line 37, in run
    exec(code, {'__name__': '__main__'})
  File "test-cli.py", line 1, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "I:\REF\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 91, in <module>
    _setupQtDirectories()
  File "I:\REF\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 57, in _setupQtDirectories
    import shiboken2
  File "<frozen zipimport>", line 259, in load_module
  File "I:\REF\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\shiboken2\__init__.py", line 27, in <module>
    from .shiboken2 import *
ImportError: DLL load failed while importing shiboken2: The specified module could not be found.

Everything works fine by directly calling the Python script as follows:

> python test-gui.py

I tried with the following version of Pyside2: 5.14.2.1, 5.14.2.2 and 5.14.2.3.

I am using the following cx_freeze configuration: [cx_freeze_conf.zip](https://github.com/marcelotduarte/cx_Freeze/files/5402159/cx_freeze_conf.zip)

Platform configuration:
- Windows 10 x86_64
- Python V3.8.3
- cx_Freeze 6.1
- PySide2 5.14.2.3

**Comment by marcelotduarte** [comment]
cx_Freeze 6.3 has just been released.
Can you check if this has been resolved?
I think this is duplicate of #665

**Comment by Lois31** [comment]
Same behavior with cx_Freeze 6.3.
The problem is not quite the same as in ticket #665 .

In issue #665 , the problem comes from using PySide2 5.14.2.1 and placing PySide2 in the zip file with the `zip_include_packages` option of cx_Freeze. This problem is fixed with PySide2 5.14.2.3

Platform configuration:

- Windows 10 x86_64
- Python V3.8.3
- cx_Freeze 6.3
- PySide2 5.14.2.3

**Comment by Lois31** [comment]
Hello Team,
Have you been able to make progress on this problem?
Thanks

**Comment by Lois31** [comment]
For information, I did some tests with different versions of PySide2:

**5.14.1**
The executable is launched but hangs indefinitely.
I have to force it to stop with a Ctrl+C and the following Python stack appears:
```
test-cli.exe -h
Traceback (most recent call last):
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\genericpath.py", line 19, in exists
    os.stat(path)
FileNotFoundError: [WinError 2] Le fichier spécifié est introuvable: '\\\\filer001\\HOME25\\files.dir\\shibokensupport\\signature\\loader.py'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "(builtin)", line 113, in bootstrap
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\genericpath.py", line 19, in exists
    os.stat(path)
KeyboardInterrupt
Fatal Python error: could not initialize part 2
Python runtime state: initialized

Current thread 0x00004390 (most recent call first):
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\shiboken2\__init__.py", line 30 in <module>
  File "<frozen zipimport>", line 259 in load_module
  File "<frozen importlib._bootstrap>", line 618 in _load_backward_compatible
  File "<frozen importlib._bootstrap>", line 655 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 975 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 991 in _find_and_load
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 25 in _setupQtDirectories
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 61 in <module>
  File "<frozen zipimport>", line 259 in load_module
  File "<frozen importlib._bootstrap>", line 618 in _load_backward_compatible
  File "<frozen importlib._bootstrap>", line 655 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 975 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 991 in _find_and_load
  File "test-gui.py", line 1 in <module>
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\Console.py", line 37 in run
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\__startup__.py", line 40 in run
```

**5.14.2.3**
Issue described in the description of this ticket.

**5.15.1**
```
test-cli.exe -h
PySide2/__init__.py: Unable to import shiboken2 from L:\dev\ALTO\bin\windows, L:\dev\ALTO\bin\windows\lib\library.zip, L:\dev\ALTO\bin\windows\lib
Traceback (most recent call last):
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\__startup__.py", line 40, in run
    module.run()
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\Console.py", line 37, in run
    exec(code, {'__name__': '__main__'})
  File "test-gui.py", line 1, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 107, in <module>
    _setupQtDirectories()
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 58, in _setupQtDirectories
    import shiboken2
  File "<frozen zipimport>", line 259, in load_module
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\shiboken2\__init__.py", line 27, in <module>
    from .shiboken2 import *
ImportError: DLL load failed while importing shiboken2: Le module spécifié est introuvable.
```

I am completely stuck on this issue. Can you help me unblock the situation?

**Comment by marcelotduarte** [comment]
@Lois31 The sample at https://github.com/marcelotduarte/cx_Freeze/tree/master/cx_Freeze/samples/pyside2 with pyside2 5.14.1 should work. Can you test it to compare?

**Comment by Lois31** [comment]
This example does not work in my case. 
As explained in my previous comment, PySide2 version 5.14.1 hangs indefinitely. 
It seems to be looking for the _\\\\filer001\\HOME25\\files.dir\\shibokensupport\\loader.py_ file, but not in the installation directory which is, in my case: `I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows`
Could you explain this behavior and especially offer me a workaround?

**Comment by klensy** [comment]
It's looks like network share ` \\filer001\...` ? You building app in system environment (not in virtual env)? Try to build it in virtualenv for test.

**Comment by Lois31** [comment]
_filer001_ is indeed a shared network drive. 
I just tested by creating a Virtual env, and I have the exact same error ...
Could you explain why shiboken is looking for the shibokensupport folder outside of the delivery directory?

**Comment by klensy** [comment]
Tested example with pyside2==5.14.2.3 (on windows)

It will work, at least run without errors (but with current algo for dependencies search it hard to do it auto).
Need to add to `lib` folder `shiboken2.abi3.dll` from shiboken2 package, plus `pyside2.abi3.dll` and Qt libs with same names, like pyd files already in (for `PySide2.QtGui.pyd` need to add `Qt5Gui.dll`) from pyside2 package.

**Comment by Lois31** [comment]
I installed PySide2 version 5.14.2.3. 
I then copy the following files to the lib directory:

- shiboken2/shiboken2.abi3.dll
- PySide2/pyside2.abi3.dll
- PySide2/Qt5*.dll

However, it still makes it impossible to load PySide2.
I have also tried adding or not the PySide2 and shiboken2 modules in the zip file, but there is no impact.

Moreover, when the PySide2 and shiboken2 modules are not in the zip file, all the files that I am asked to copy are in the respective folders:
- lib/shiboken2
- lib/PySide2

To finish, I test to copy all of the PySide2 module files in the lib/shiboken2 folder and I get a different error:
```
PySide2/__init__.py: Unable to import shiboken2 from L:\dev\ALTO\bin\windows, L:\dev\ALTO\bin\windows\lib\library.zip, L:\dev\ALTO\bin\windows\lib
Traceback (most recent call last):
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\__startup__.py", line 40, in run
    module.run()
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\Console.py", line 37, in run
    exec(code, {'__name__': '__main__'})
  File "test-gui.py", line 1, in <module>
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 91, in <module>
    _setupQtDirectories()
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 57, in _setupQtDirectories
    import shiboken2
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 91, in <module>
    _setupQtDirectories()
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 65, in _setupQtDirectories
    type.__signature__
AttributeError: type object 'type' has no attribute '__signature__'
```
Maybe this can help in solving the problem

**Comment by Lois31** [comment]
I also tested copying the PySide2 and shiboken2 modules directly to the installation directory. And weirdly, I get the same error as when cx_Freeze does the job:
```
PySide2/__init__.py: Unable to import shiboken2 from L:\dev\ALTO\bin\windows, L:\dev\ALTO\bin\windows\lib\library.zip, L:\dev\ALTO\bin\windows\lib
Traceback (most recent call last):
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\__startup__.py", line 40, in run
    module.run()
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\Console.py", line 37, in run
    exec(code, {'__name__': '__main__'})
  File "test-gui.py", line 1, in <module>
  File "L:\dev\ALTO\bin\windows\PySide2\__init__.py", line 91, in <module>
    _setupQtDirectories()
  File "L:\dev\ALTO\bin\windows\PySide2\__init__.py", line 57, in _setupQtDirectories
    import shiboken2
  File "L:\dev\ALTO\bin\windows\shiboken2\__init__.py", line 27, in <module>
    from .shiboken2 import *
ImportError: DLL load failed while importing shiboken2: Le module spécifié est introuvable.
```
Are you sure there are no missing libraries to import for PySide2 to work properly?

**Comment by klensy** [comment]
Oh, there also dependency on Qt5Qml.dll, my sorry.
listing for files for pyside2 example:
```
.:
./  ../  lib/  platforms/  PySide2app.exe*  python3.dll*  python38.dll*  vcruntime140.dll*

./lib:
./             collections/       PySide2.QtCore.pyd*     Qt5Gui.dll*      shiboken2.abi3.dll*
../            importlib/         PySide2.QtGui.pyd*      Qt5Network.dll*  shiboken2.shiboken2.pyd*
_bz2.pyd*      library.zip        PySide2.QtNetwork.pyd*  Qt5Qml.dll*      unicodedata.pyd*
_hashlib.pyd*  logging/           PySide2.QtWidgets.pyd*  Qt5Widgets.dll*  urllib/
_lzma.pyd*     pyside2.abi3.dll*  Qt5Core.dll*            select.pyd*

./lib/collections:
./  ../  __init__.pyc  abc.pyc

./lib/importlib:
./   __init__.pyc    _bootstrap_external.pyc  machinery.pyc  util.pyc
../  _bootstrap.pyc  abc.pyc                  metadata.pyc

./lib/logging:
./  ../  __init__.pyc

./lib/urllib:
./  ../  __init__.pyc  parse.pyc

./platforms:
./  ../  qdirect2d.dll*  qminimal.dll*  qoffscreen.dll*  qwebgl.dll*  qwindows.dll*
```

**Comment by Lois31** [comment]
Finally a breakthrough! Thanks for the help. 
However, the window still does not launch correctly ...

Let me explain:
First of all, after having feezed the [example](https://github.com/marcelotduarte/cx_Freeze/tree/master/cx_Freeze/samples/pyside2), with 
 - Python 3.8.3
 - cx_Freeze 6.1
 - PySide2 5.14.2.3 

I get the following tree:
```
.:
./  ../  lib/  platforms/  PySide2app.exe*  python38.dll*

./lib:
./             collections/       PySide2.QtCore.pyd*
../            importlib/         PySide2.QtGui.pyd*      shiboken2.shiboken2.pyd*
_bz2.pyd*      library.zip        PySide2.QtNetwork.pyd*  unicodedata.pyd*
_hashlib.pyd*  logging/           PySide2.QtWidgets.pyd*  urllib/
_lzma.pyd*      select.pyd*

./lib/collections:
./  ../  __init__.pyc  abc.pyc

./lib/importlib:
./   __init__.pyc    _bootstrap_external.pyc  machinery.pyc  util.pyc
../  _bootstrap.pyc  abc.pyc                  metadata.pyc

./lib/logging:
./  ../  __init__.pyc

./lib/urllib:
./  ../  __init__.pyc  parse.pyc

./platforms:
./  ../  qdirect2d.dll*  qminimal.dll*  qoffscreen.dll*  qwebgl.dll*  qwindows.dll*
```

To avoid having an error when launching, I have to copy the following files:
- python3.dll
- vcruntime140.dll
- lib/pyside2.abi3.dll
- lib/Qt5Core.dll
- lib/Qt5Gui.dll
- lib/Qt5Network.dll
- lib/Qt5Widgets.dll
- lib/shiboken2.abi3.dll

Once these libs are copied, the shiboken2 import error disappears. **BUT no window appears** ...
After forcing the executable to stop, below is the Python stack:
```
Traceback (most recent call last):
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\genericpath.py", line 19, in exists
    os.stat(path)
FileNotFoundError: [WinError 2] Le fichier spécifié est introuvable: '\\\\filer001\\HOME25\\files.dir\\shibokensupport\\signature\\loader.py'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "(builtin)", line 113, in bootstrap
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\genericpath.py", line 19, in exists
    os.stat(path)
KeyboardInterrupt
Fatal Python error: could not initialize part 2
Python runtime state: initialized

Current thread 0x00000518 (most recent call first):
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\shiboken2\__init__.py", line 30 in <module>
  File "<frozen zipimport>", line 259 in load_module
  File "<frozen importlib._bootstrap>", line 618 in _load_backward_compatible
  File "<frozen importlib._bootstrap>", line 655 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 975 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 991 in _find_and_load
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 57 in _setupQtDirectories
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\PySide2\__init__.py", line 91 in <module>
  File "<frozen zipimport>", line 259 in load_module
  File "<frozen importlib._bootstrap>", line 618 in _load_backward_compatible
  File "<frozen importlib._bootstrap>", line 655 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 975 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 991 in _find_and_load
  File "test-gui.py", line 1 in <module>
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\Console.py", line 37 in run
  File "I:\DEV\COTS\PYTHON\V3.8.3\BINARIES\Windows\lib\site-packages\cx_Freeze\initscripts\__startup__.py", line 40 in run
```

Do you have an idea ?

**Comment by klensy** [comment]
>cx_Freeze 6.1

Why you used outdated version? Try 6.3 version.

**Comment by marcelotduarte** [comment]
@Lois31 cx_Freeze 6.1 has issues with shiboken, that are resolved since 6.2. Please use 6.3

**Comment by Lois31** [comment]
I just installed cx_Freeze 6.3. 
After having copied the missing Qt5 libs, the executable launches without error **but blocks indefinitely**, as explained in the ticket above.

It seems to be blocking, still on the same file,  `files.dir\\shibokensupport\\signature\\loader.py`
```
FileNotFoundError: [WinError 2] The specified file could not be found: '\\\\ filer001 \\ HOME25 \\ files.dir \\ shibokensupport \\ signature \\ loader.py'
```

**Comment by marcelotduarte** [comment]
@Lois31 This topic is getting big, so it is sometimes difficult to follow, so I will summarize:
- Create a new virtual environment (python -m venv ...)
- Use cx-freeze 6.3 (pip install cx-freeze==6.3)
- Install pyside2 5.14.1 (pip install pyside2==5.14.1)
- Test the sample at https://github.com/marcelotduarte/cx_Freeze/tree/master/cx_Freeze/samples/pyside2
- Put the full traceback to me
- If you can, install the lasted python 3.8.6 in a local C drive and do the same test

**Comment by Lois31** [comment]
I am starting to understand where the problem comes from after doing some testing.
Whether I'm using my Python version 3.8.3 (installed on a filer), a virtual environment, or even a newer version of Python 3.8.6 (installed in C:\Program Files), the problem is always the same:
The executable `PySide2app.exe` launches without error but no window appears. Once we launch a kill of the process, we can see that it blocks on the loading of the file: `\\\\filer001\\HOME25\\files.dir\\shibokensupport\\signature\\loader.py`
Note that the name `\\filer001\\HOME 25\` represents my workspace on a filer (network hard drive).`

So I moved my source code from my workspace (`\\filer011\HOME25`) to the local hard drive of the machine (`C:\Temp`) and start the compilation with the new version of Python 3.8.6 (installed in C:\Program Files).
Result: the window is launched and appears as expected!
**I restarted the compilation with my Python version 3.8.3 (installed on a filer). Identical result: the window is launched and appears as expected!**

So I ask myself the question: is there a problem in the management of the UNC path between cx_Freeze and PySide2?

**Comment by klensy** [comment]
Don't launch anything from network share, it will cause a lot of errors for a lot of programs.

**Comment by Lois31** [comment]
Last observation, the functional executable in `C:\Temp` no longer works by moving it to my workspace (`\\filer011\HOME25`)

Except that for us, this is not an option for infrastructure reasons ...

**Comment by Lois31** [comment]
Does cx_Freeze modify the `shiboken2/shiboken2.abi3.dll` library at the path level?
Or any other modification that could cause this problem?

**Comment by marcelotduarte** [comment]
But the sample, you tested with it? I insist on the sample because it is minimal. It is a start point.

I'll tell you about an experience with network errors. I sometimes test some samples before releasing a version and some did not run. My base machine is Linux and accessed from a machine via samba, and ran 2 tests, using MSYS2 to compile and then the native python. When I copied the samples to the local machine, the problems stopped, so I attributed this problem to samba / msys, since in the native python version it was no problem.
Another error I saw apparently was with the msys version pip, since it gave an error before running cx-freeze. These errors in msys started when they switched to py38, before with py37 it hadn't.
Now I see you with errors with native py38.

**Comment by marcelotduarte** [comment]
> is there a problem in the management of the UNC path between cx_Freeze and PySide2?

I think it is in package management - pip, setuptools, etc.

> Does cx_Freeze modify the shiboken2/shiboken2.abi3.dll library at the path level?

cx-freeze doesn't change much on windows, in dll it doesn't change anything. What happens is the way windows searches for dlls, and this may be affecting them. In windows, the search for dlls is in the path, and the search order in the path may be affecting finding those dlls.
You may have to change this when in freeze mode: https://docs.python.org/3.8/library/os.html#os.add_dll_directory

**Comment by Lois31** [comment]
I tested adding the `os.add_dll_directory` function as follows:
```
import os
import sys
if getattr(sys, 'frozen', False):
    dll_path = os.path.join(os.path.dirname(sys.executable), 'lib')
    print(dll_path)
    os.add_dll_directory(dll_path)

from PySide2 import QtGui, QtCore, QtUiTools, QtWidgets, QtXml

print("Import OK")
```
But nothing changes ...

What I don't understand is that when my application is not frozen, it works fine, even on filer.
How can cx_Freeze change this behavior?

**Comment by marcelotduarte** [comment]
cx_Freeze 6.6 has just been released.
Assuming this has been resolved.
If you had issues please open a new issue.
